### PR TITLE
Re-Implemented ILBinaryReader with System.Reflection.Metadata

### PR DESCRIPTION
**Pull request by TIHan**
This work is almost a year old, Jan 4th 2019; I decided to go ahead and try to finish it since I've been doing a lot of work in our IL metadata lately.

It saves about ~1500 lines of code and is a lot more readable.

**Acceptance Criteria**
- [ ] Merge with latest 'main'
- [ ] Properly read field data as a byte array when the field has a RVA
- [ ] Add a command-line flag to turn this the reader on as we need to keep the original metadata reader available in-case of regressions; means this will be opt-in for now until we feel very comfortable that it works well enough to be the default.
- [ ] Manual testing of major F# open-source projects, FAKE, Ionide, etc.
- [ ] Tests that verify same data that was read in the older metadata reader

### CODE REVIEW

File: src/absil/il.fs [hunk 1]
```diff
@@ -3950,10 +3962,11 @@ let decodeILAttribData (ilg: ILGlobals) (ca: ILAttribute) =
             parseElems (v :: acc) (n-1) sigptr
           let elems, sigptr = parseElems [] n sigptr
           ILAttribElem.Array (elemTy, elems), sigptr
+      | ILType.Boxed _
```
Comment (TIHan) [inline src/absil/il.fs]: Should probably remove this since I might have fixed it.

**Comment by TIHan** [comment]
This is almost passing CI. Having issues when it comes to using `--standalone` and possible `--staticlink`.

**Comment by TIHan** [comment]
@dsyme this is a big one and am happy to go over it with you.

One thing I want to make sure is that `--standalone` and `--staticlink` work as before. I am creating separate PRs that add more tests to those features.
I also want to write tests that truly ensure I have the correct `ILType.TypeVar` index, especially for multiple deep nested types and their method signatures.

**Comment by TIHan** [comment]
I want to consider removing `tryGetMetadataSnapshot` in this PR.

I know the purpose of it is to use the raw pointer to the metadata given from VS, but the raw pointer isn't the pointer to the actual PE, just the pointer to where the metadata is inside of the PE. Our `ILModuleReader` requires data from the PE which the raw pointer can't give us. So, we have to open the PE file on-disk to read info it only needs and then opt to use the raw pointer given by `tryGetMetadataSnapshot`. That's not completely safe as it is technically possible to have the states of what is read from disk and given from `tryGetMetadataSnapshot` to be incompatible.

While it is better to use the raw pointer given from VS, I think we would be ok if we simply used a memory-mapped file that shadow-copies the assembly. This will increase memory usage on the logical address space, but won't impact the IDE processes themselves to cause any out-of-memory problems.

File: src/absil/ilread.fs [hunk 2]
```diff
@@ -1,3939 +1,2582 @@
...
-    lazy
-        let (ctxt: ILMetadataReader) = getHole ctxtH
-        let mdv = ctxt.mdfile.GetView()
-        let dataStartPoints = 
-            let res = ref []
-            for i = 1 to ctxt.getNumRows TableNames.FieldRVA do
-                let rva, _fidx = seekReadFieldRVARow ctxt mdv i
-                res := ("field", rva) :: !res
-            for i = 1 to ctxt.getNumRows TableNames.ManifestResource do
-
```
Comment (TIHan) [inline src/absil/ilread.fs]: I may not actually need this, but it could replace `decodeILAttribData` in `il.fs`.

Comment (TIHan) [inline src/absil/ilread.fs]: Let's not read the entire block, should just use an instance of `ByteMemory` that spans over the specific content.

File: src/absil/il.fsi [hunk 3]
```diff
@@ -1815,6 +1818,7 @@ val mkILCustomAttrsFromArray: ILAttribute[] -> ILAttributes
 val storeILCustomAttrs: ILAttributes -> ILAttributesStored
 val mkILCustomAttrsReader: (int32 -> ILAttribute[]) -> ILAttributesStored
 val emptyILCustomAttrs: ILAttributes
+val emptyILCustomAttrsStored: ILAttributesStored
```
Comment (dsyme) [inline src/absil/il.fsi]: Might be good to put all these smaller changes in a separate PR

File: src/absil/ilread.fsi [hunk 4]
```diff
@@ -24,84 +24,86 @@
 ///     class.  That is not particularly satisfactory, and it may be
 ///     a good idea to build a small library which extracts the information
 ///     you need.  
-module FSharp.Compiler.AbstractIL.ILBinaryReader
```
Comment (dsyme) [inline src/absil/ilread.fsi]: I think the changes in this file could be minimized by not changing `module` to `namespace` at the top.

File: src/absil/il.fsi (hunk 3 above)
Comment (TIHan) [reply src/absil/il.fsi]: Yea that makes sense. Less changes outside `ilread.fs` the better.

File: src/absil/ilread.fsi (hunk 4 above)
Comment (TIHan) [reply src/absil/ilread.fsi]: Yea, I see what I need to do in order to preserve it.

**Comment by TIHan** [comment]
I agree, definitely agree a more close review should be done once everything is passing because who knows what changes would have to be done :).

I'm doing what I can to keep everything as isolated as possible.

**Comment by TIHan** [comment]
Wow, CI passes for the first time.

However, there is a perf issue when getting the content of the field using the field's RVA. I think it's reading more than it should.

**Comment by cartermp** [comment]
Are there any interesting benchmark results now that it's passing CI?

File: src/absil/ilread.fs [hunk 5]
```diff
@@ -1,4005 +1,2561 @@
...
+            InlineNone(volatileOrUnalignedPrefix (fun (ilAlignment, ilVolatility) -> I_ldind(ilAlignment, ilVolatility, DT_U2)))//byte OperandType.InlineNone           // ldind.u2
+            InlineNone(volatileOrUnalignedPrefix (fun (ilAlignment, ilVolatility) -> I_ldind(ilAlignment, ilVolatility, DT_I4)))//byte OperandType.InlineNone           // ldind.i4
+            InlineNone(volatileOrUnalignedPrefix (fun (ilAlignment, ilVolatility) -> I_ldind(ilAlignment, ilVolatility, DT_U4)))//byte OperandType.InlineNone           // ldind.u4
+            InlineNone(volatileOrUnalignedPrefix (fun (ilAlignment, ilVolatility) -> I_ldind(ilAlignment, ilVolatility, DT_I8)))//byte OperandType.InlineNone           // ldind.i8
+            InlineNone(volatileOrUnalignedPrefix (fun (ilAlignment, ilVolatility) -> I_ldind(ilAlignment, ilVolatility, DT_I)))//byte OperandType.InlineNone           // ldind.i
+            InlineNone(volatileOrUnalignedPrefix (fun (ilAlignment, ilVolatility) -> I_ldind(ilAlignment, ilVolatility, DT_R4)))//byte OperandType.InlineNone           // ldind.r4
+
```
Comment (TIHan) [inline src/absil/ilread.fs:1972]: This is incorrect: ```blobReader.ReadBytes(blobReader.Length)```

`blobReader.Length` will definitely not give back the proper length. We may have to do what the original implementation did.

**Comment by vzarytovskii** [comment]
Shelving this for the time being.
